### PR TITLE
many: fix test cases to work with different DistroLibExecDir

### PR DESCRIPTION
**Pull request by morphis**
This is needed to allow execution of our unit tests within the package build on other distributions like Fedora.

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -90,7 +91,7 @@ func ExecInCoreSnap() {
 	// ensure we do not re-exec into an older version of snapd, look
 	// for info file and ignore version of core that do not yet have
 	// it
-	fullInfo := filepath.Join(corePath, "/usr/lib/snapd/info")
```
Comment (zyga) [inline cmd/cmd.go]: Wait, this feels wrong. If core path is the path where core is mounted the rest of the string can be a constant because it's not depending on the distribution anymore.

File: cmd/snap/cmd_run_test.go [hunk 2]
```diff
@@ -449,8 +450,8 @@ func (s *SnapSuite) TestSnapRunIsReexeced(c *check.C) {
 		readlink string
 		expected bool
 	}{
-		{filepath.Join(dirs.SnapMountDir, "/usr/lib/snapd/snapd"), true},
-		{"/usr/lib/snapd/snapd", false},
+		{filepath.Join(dirs.SnapMountDir, dirs.DistroLibExecDir, "snapd"), true},
```
Comment (zyga) [inline cmd/snap/cmd_run_test.go]: Again, isn't this constant?

File: cmd/cmd.go (hunk 1 above)
Comment (morphis) [inline cmd/cmd.go]: I will into this again. However all tests except refresh-delta-from-core on Ubuntu 14.04 seem to pass.

File: cmd/snap/cmd_run_test.go (hunk 2 above)
Comment (morphis) [inline cmd/snap/cmd_run_test.go]: For now yes, but when we have switch to different core/base snaps it may not. Using dirs.CoreLibExecDir here now.

Comment (zyga) [inline cmd/snap/cmd_run_test.go]: using `CoreLibExecDir` is fine. For core/base split I don't know how it will work yet so it's hard to say.

File: cmd/cmd.go (hunk 1 above)
Comment (zyga) [inline cmd/cmd.go]: This looks correct now.

File: cmd/snap/main_test.go [hunk 3]
```diff
@@ -152,9 +152,13 @@ func mockVersion(v string) (restore func()) {
 	return func() { cmd.Version = old }
 }
 
-func mockSnapConfine() func() {
-	snapConfine := filepath.Join(dirs.DistroLibExecDir, "snap-confine")
-	if err := os.MkdirAll(dirs.DistroLibExecDir, 0755); err != nil {
+func mockSnapConfine(reexec bool) func() {
```
Comment (zyga) [inline cmd/snap/main_test.go]: Just nitpick. I'd prefer a flag like ReExecEnabled or something similar. The false/true that is in tests above is hard to grok just by looking.

Comment (morphis) [inline cmd/snap/main_test.go]: Let me change that.

File: cmd/snap/cmd_run.go [hunk 4]
```diff
@@ -228,7 +228,11 @@ func isReexeced() bool {
 }
 
 func runSnapConfine(info *snap.Info, securityTag, snapApp, command, hook string, args []string) error {
-	snapConfine := filepath.Join(dirs.DistroLibExecDir, "snap-confine")
+	libExecDir := dirs.DistroLibExecDir
+	if isReexeced() {
```
Comment (niemeyer) [inline cmd/snap/cmd_run.go]: This should be done only once in this function. Right now it's being done twice, and it's also bogus because code right below will test for a path that does not exist. Proper way is likely to bring the code currently at the bottom here, and make sure snapConfine points to the proper thing upfront.

File: cmd/snap/main_test.go (hunk 3 above)
Comment (niemeyer) [inline cmd/snap/main_test.go]: Or even simpler, pass libExecDir as a parameter.

**Comment by Conan-Kudo** [comment]
I'm confused... Why are so many things changing from DistroLibExecDir to CoreLibExecDir? That seems to indicate it'd be running/testing a completely different/wrong thing...

**Comment by morphis** [comment]
@Conan-Kudo Many of the changes are for paths are in the code path for the re-execution feature which never worked on Fedora/Suse. So you never saw anything going wrong. This PR fixes those cases and brings us a bit closer to working re-execution too.

**Comment by mvo5** [comment]
Looks good, but some review comments (like https://github.com/snapcore/snapd/pull/3222#discussion_r113169155 are not addressed yet)

File: cmd/snap/main_test.go (hunk 3 above)
Comment (morphis) [inline cmd/snap/main_test.go]: Using libExecDir now.

File: cmd/snap/cmd_run.go (hunk 4 above)
Comment (morphis) [inline cmd/snap/cmd_run.go]: Reworked the logic a bit.

**Comment by morphis** [comment]
Fixed all review comments and rebased on latest master

File: cmd/cmd.go (hunk 1 above)
Comment (zyga) [inline cmd/cmd.go]: Wait, this is not what I expected (the diff looked different before). So on Fedora the core path is `/var/lib/snapd/snap/core/current/` but then the `dirs.DistroLibExecDir` will contain `/usr/libexec/snapd`, I don't understand how this is supposed to work.

**Comment by mvo5** [comment]
Some unit test failures, e.g. https://travis-ci.org/snapcore/snapd/builds/226369117#L2580

File: cmd/cmd.go (hunk 1 above)
Comment (morphis) [inline cmd/cmd.go]: @zyga Please note that the code now uses dirs.CoreLibExecDir which will expand to /usr/lib/snapd even on Fedora. A previous version of the code was using dirs.DistroLibExecDir which is, as you pointed out, wrong.

In summary: The new code `filepath.Join(corePath, dirs.CoreLibExecDir, "info")` will take the `corePath` as determined before through `/proc/self/exe` and append dirs.CoreLibExecDir so that we end up with `/snap/core/current/usr/libexec/snapd/info` on Ubuntu and `/var/lib/snapd/snap/core/current/usr/libexec/snapd/info`

**Comment by morphis** [comment]
@mvo5 Pushed a change to fix the failing unit tests.

File: interfaces/builtin/content_test.go [hunk 5]
```diff
@@ -195,11 +199,11 @@ apps:
 
 func (s *ContentSuite) TestResolveSpecialVariable(c *C) {
 	info := snaptest.MockInfo(c, "name: name", &snap.SideInfo{Revision: snap.R(42)})
-	c.Check(builtin.ResolveSpecialVariable("foo", info), Equals, "/snap/name/42/foo")
```
Comment (zyga) [inline interfaces/builtin/content_test.go]: Sorry for taking so long to verify this. This part is incorrect. The generated mount profile is only processed after pivot_root has happened so it must use the internal layout.

Comment (morphis) [inline interfaces/builtin/content_test.go]: How that? On Debian & Ubuntu we have /snap and on Fedora we have /var/lib/snapd/snap in and outide of the snap or do I miss anything?

Comment (zyga) [inline interfaces/builtin/content_test.go]: You are missing one thing. After pivot_root whatever the snap mount directory is it ends up being /snap. Only then do we process those mount directives so the hard-coded paths are actually always correct.

Comment (morphis) [inline interfaces/builtin/content_test.go]: Running the following

```
$ snap run --shell hello-world
bash-4.3$ echo $SNAP
/var/lib/snapd/snap/hello-world/27
```

Please also note that I didn't change builting.ResolveSpecialVariable. It is using dirs.BaseSnapMountDir already without this PR. See https://github.com/snapcore/snapd/blob/master/interfaces/builtin/content.go#L123

Comment (zyga) [inline interfaces/builtin/content_test.go]: As discussed on IRC there are a host of bugs here:

- snap execution environment should look the same everywhere
- $SNAP is always /snap/$SNAP_NAME/$SNAP_REVISION when seen from inside snap execution environment
- various other things are wrong but we'll discuss and fix them in separate PRs

File: tests/unit/go/task.yaml [hunk 6]
```diff
@@ -4,9 +4,18 @@ systems: [ubuntu-16.04-64]
...
+    rm -f /etc/os-release
+    mv /etc/os-release.orig /etc/os-release
 
 execute: |
     mkdir -p /tmp/unit-tests/src/github.com/snapcore
     cp -ar $PROJECT_PATH /tmp/unit-tests/src/github.com/snapcore
     chown -R test:12345 /tmp/unit-tests
     su -l -c "cd /tmp/unit-tests/src/github.com/snapcore/snapd && GOPATH=/tmp/unit-tests ./run-checks --unit" test
+
+    mv /etc/os-release /etc/os-release.orig
```
Comment (zyga) [inline tests/unit/go/task.yaml]: This is more of a `prepare:` stage to me but the actual section looks curious. Is this intentional?

Comment (morphis) [inline tests/unit/go/task.yaml]: Fixed and reworked the test case a bit more.

File: interfaces/builtin/content_test.go (hunk 5 above)
Comment (morphis) [inline interfaces/builtin/content_test.go]: The code is now changed to use /snap as static prefix.

File: dirs/dirs.go [hunk 7]
```diff
@@ -126,9 +128,11 @@ func SetRootDir(rootdir string) {
 
 	switch release.ReleaseInfo.ID {
 	case "fedora", "centos", "rhel", "arch":
-		SnapMountDir = filepath.Join(rootdir, "/var/lib/snapd/snap")
+		BaseSnapMountDir = "/var/lib/snapd/snap"
```
Comment (niemeyer) [inline dirs/dirs.go]: We can't use the "BaseSnap" term here, even more when we have a matching variable with "CoreSnap". Base snaps are around the block and this is unrelated.

File: dirs/dirs.go [hunk 8]
```diff
@@ -179,7 +183,8 @@ func SetRootDir(rootdir string) {
 		DistroLibExecDir = filepath.Join(rootdir, "/usr/lib/snapd")
 	}
 
-	CoreLibExecDir = filepath.Join(rootdir, "/usr/lib/snapd")
+	CoreLibExecDir = "/usr/lib/snapd"
+	CoreSnapMountDir = "/snap"
```
Comment (niemeyer) [inline dirs/dirs.go]: What's the line for deciding what has a prefix or not?  Why do we have BaseSnapMountDir with rootdir, but the ones above not?  Hard to guess on call sites, which makes it quite error prone.

File: dirs/dirs.go (hunk 7 above)
Comment (Conan-Kudo) [inline dirs/dirs.go]: `HostSnapMountDir` would make more sense, ne?

File: dirs/dirs.go (hunk 8 above)
Comment (morphis) [inline dirs/dirs.go]: We need a variable which has just the clean mount dir without the root prefix as otherwise we end up with something $prefix/$prefix/$snapmountdir in the unit tests. See https://github.com/snapcore/snapd/pull/3222/files#diff-76f691a17e89c1330742e51bb468ef76R203 as an example. The only place where is ever being used are unit tests. I can move these variables into test code and keep it away from dirs/dirs.go actually would be nice to keep these things in a central place.

Also CoreLibExecDir are needed as these describe where in the core snap those directories are located. These are different on Fedora than the outside SnapMountDir/LibExecDir so we have to keep them distinct.

File: dirs/dirs.go (hunk 7 above)
Comment (morphis) [inline dirs/dirs.go]: `HostSnapMountDir` still has a notion of not saying that it is absolute or not. `BaseSnapMountDir` was my attempt to word that. What about `RawSnapMountDir`?

File: dirs/dirs.go (hunk 8 above)
Comment (pedronis) [inline dirs/dirs.go]: if a unit test tries to write to those Core* dirs missing the rootdir will creates errors

File: dirs/dirs.go [hunk 9]
```diff
@@ -32,7 +32,18 @@ import (
...
+	RawSnapMountDir string
+	// CoreSnapMountDir is the path of the snap mount dir inside
+	// the snap confinement environment. Give this nature it will
+	// not include the GlobalRootDir as prefix.
+	CoreSnapMountDir string
+	// CoreLibExecDir is similar to CoreSnapMountDir but for the
+	// LibExecDir used by snapd.
+	CoreLibExecDir            string
+	DistroLibExecDir          string
 	SnapBlobDir               string
```
Comment (pedronis) [inline dirs/dirs.go]: we probably want a newline to separate things here

Comment (morphis) [inline dirs/dirs.go]: Will add one.

File: dirs/dirs.go (hunk 8 above)
Comment (morphis) [inline dirs/dirs.go]: The unit test doesn't try to write into those dirs. They are used for example in the content interface where we check if the mount entries generated are ok. @zyga told me that those should never use SnapMountDir (which might be prefixed) but the mount dir without any prefix so /snap on Ubuntu and /var/lib/snapd/snap on Fedora.

Comment (niemeyer) [inline dirs/dirs.go]: I suggest keeping the rootdir as originally, and stripping GlobalRootDir out of it when really needed. That makes it consistent.

File: dirs/dirs.go (hunk 7 above)
Comment (niemeyer) [inline dirs/dirs.go]: That makes it Raw, except other things raw are not Raw. Still pretty misleading and error prone. Not to mention "raw" doesn't really say much in this context.

See the suggestion in the other point.

File: dirs/dirs.go (hunk 8 above)
Comment (morphis) [inline dirs/dirs.go]: @niemeyer Doing that now. Please have another look.

File: dirs/dirs.go [hunk 10]
```diff
@@ -32,7 +32,18 @@ import (
 var (
 	GlobalRootDir string
 
-	SnapMountDir              string
+	SnapMountDir string
+
+	// CoreSnapMountDir is the path of the snap mount dir inside
+	// the snap confinement environment. Give this nature it will
```
Comment (zyga) [inline dirs/dirs.go]: `Given its nature` maybe?

File: interfaces/builtin/content.go [hunk 11]
```diff
@@ -116,13 +117,23 @@ func (iface *contentInterface) path(slot *interfaces.Slot, name string) []string
 	return out
 }
 
+const (
+	SnapMountDir = iota
+	SnapDataDir
+	SnapCommonDir
```
Comment (zyga) [inline interfaces/builtin/content.go]: Wait, are those even used?

Comment (morphis) [inline interfaces/builtin/content.go]: Hm, they aren't. Seems to be a leftover. Good spot!

Comment (morphis) [inline interfaces/builtin/content.go]: Drop these now.

**Comment by morphis** [comment]
@pedronis @zyga @Conan-Kudo @niemeyer Can you have another look so we get this reviewed and merged soon?

File: cmd/snap/cmd_run_test.go [hunk 12]
```diff
@@ -98,7 +98,7 @@ func (s *SnapSuite) TestSnapRunAppIntegration(c *check.C) {
 	// mock installed snap
 	dirs.SetRootDir(c.MkDir())
 	defer func() { dirs.SetRootDir("/") }()
-	defer mockSnapConfine()()
+	defer mockSnapConfine(dirs.DistroLibExecDir)()
```
Comment (mvo5) [inline cmd/snap/cmd_run_test.go]: Pardon my ignorance, but it looks like the argument for `mockSnapConfine()` is always dirs.DistroLibExecDir - do we really need this new argument in this case?

File: cmd/snap/cmd_run_test.go [hunk 13]
```diff
@@ -451,8 +451,8 @@ func (s *SnapSuite) TestSnapRunIsReexeced(c *check.C) {
 		readlink string
 		expected bool
 	}{
-		{filepath.Join(dirs.SnapMountDir, "/usr/lib/snapd/snapd"), true},
-		{"/usr/lib/snapd/snapd", false},
+		{filepath.Join(dirs.SnapMountDir, dirs.CoreLibExecDir, "snapd"), true},
+		{fmt.Sprintf("%s/snapd", dirs.DistroLibExecDir), false},
```
Comment (mvo5) [inline cmd/snap/cmd_run_test.go]: filepath.Join() is probably more natural here

File: interfaces/builtin/content_test.go (hunk 5 above)
Comment (mvo5) [inline interfaces/builtin/content_test.go]: Silly question - if the code got changed to use /snap as static prefix, do we still need to change:
```
- "/snap/foo"
+ filepath.Join(dirs.CoreSnapMountDir, "foo")
```
I guess the answer is yes to make it more obvious when we mean "context-is-inside-core" and "context-is-outside"(?)

File: overlord/snapstate/snapstate_test.go [hunk 14]
```diff
@@ -2689,7 +2719,7 @@ version: 1.0`)
 		{
 			// ensure only local install was run, i.e. first action is pseudo-action current
 			op:  "current",
-			old: "/snap/mock/100001",
+			old: filepath.Join(dirs.StripRootDir(dirs.SnapMountDir), "mock/100001"),
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go]: This feels a bit repetitive and relatively long, I wonder if we could compute the dirs only once at the beginning of the test. OTOH it was repetitive before so maybe that is something for a later branch.

File: overlord/snapstate/snapstate_test.go [hunk 15]
```diff
@@ -4987,6 +5017,9 @@ func (s *snapmgrTestSuite) TestTrySetsTryModeJailMode(c *C) {
 	s.testTrySetsTryMode(snapstate.Flags{JailMode: true}, c)
 }
 func (s *snapmgrTestSuite) TestTrySetsTryModeClassic(c *C) {
+	if !dirs.SupportsClassicConfinement() {
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go]: Would it be possible to move everything that needs classic confinement into a single snapmgrWithClassicConfinementTestSuite and simply skip the entire suite on the above condition?

File: overlord/snapstate/snapstate_test.go [hunk 16]
```diff
@@ -5046,6 +5082,9 @@ func (s *snapmgrTestSuite) TestTryUndoRemovesTryFlagLeavesJailMode(c *C) {
 	s.testTrySetsTryMode(snapstate.Flags{JailMode: true}, c)
 }
 func (s *snapmgrTestSuite) TestTryUndoRemovesTryFlagLeavesClassic(c *C) {
+	if !dirs.SupportsClassicConfinement() {
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go]: Slightly strange that `SupportsClassicConfinement()` is a member of `dirs` - I understand why but still feels wrong. Anyway, not something to be concerned about for this branch.

File: cmd/snap/cmd_run_test.go (hunk 12 above)
Comment (morphis) [inline cmd/snap/cmd_run_test.go]: There is exactly one case where it isn't. See https://github.com/snapcore/snapd/pull/3222/files/f78d7918f26f71239000599fd11db988cd605cc0#diff-78f744cbc3eb83652f8ed773dceac016R466

File: cmd/snap/cmd_run_test.go (hunk 13 above)
Comment (morphis) [inline cmd/snap/cmd_run_test.go]: Fixed.

File: interfaces/builtin/content_test.go (hunk 5 above)
Comment (morphis) [inline interfaces/builtin/content_test.go]: I would say, we should do it this way, With that we always have a clear bind to the context and know what we're dealing with.

File: overlord/snapstate/snapstate_test.go (hunk 15 above)
Comment (morphis) [inline overlord/snapstate/snapstate_test.go]: Not sure, something I would have to look into but would prefer to do it with a follow up PR to not put more burden on this one.

File: overlord/snapstate/snapstate_test.go (hunk 14 above)
Comment (morphis) [inline overlord/snapstate/snapstate_test.go]: +1 (change + later branch)

File: cmd/snap/cmd_run_test.go (hunk 12 above)
Comment (mvo5) [inline cmd/snap/cmd_run_test.go]: Aha, silly me. I even found this in subsequent reading but then forgot to delete the comment. Sorry for the noise.

File: snap/info.go [hunk 17]
```diff
@@ -82,7 +82,12 @@ func MinimalPlaceInfo(name string, revision Revision) PlaceInfo {
 
 // MountDir returns the base directory where it gets mounted of the snap with the given name and revision.
 func MountDir(name string, revision Revision) string {
-	return filepath.Join(dirs.SnapMountDir, name, revision.String())
+	return MountDirWithBasePath(dirs.SnapMountDir, name, revision)
+}
+
+// MountDirWithBasePath returns the base directory where it gets mounted of the snap with the given name and revision.
+func MountDirWithBasePath(basePath, name string, revision Revision) string {
```
Comment (niemeyer) [inline snap/info.go]: Let's please drop this function. We don't need a special (and long) name for filepath.Join. :-)

Comment (morphis) [inline snap/info.go]: Done.
